syntax = "proto3";

package skyvault;

import "google/protobuf/empty.proto";

// WRITER
service WriterService {
  rpc WriteBatch (WriteBatchRequest) returns (WriteBatchResponse);
}

message WriteBatchRequest {
  repeated TableWriteBatchRequest tables = 1;
}

message TableWriteBatchRequest {
  string table_name = 1;
  repeated WriteBatchItem items = 2;
}

message WriteBatchItem {
  string key = 1;
  oneof operation {
    bytes value = 2;
  }
}

message WriteBatchResponse {}

// CACHE
service CacheService {
  rpc GetFromRun (GetFromRunRequest) returns (GetFromRunResponse);
}

message GetFromRunRequest {
  repeated string run_ids = 1;
  repeated string keys = 2;
}

message GetFromRunResponse {
  repeated GetFromRunItem items = 1;
}

message GetFromRunItem {
  string key = 1;
  oneof result {
    bytes value = 2;
    google.protobuf.Empty deleted = 3;
  }
}

// READER
service ReaderService {
  rpc GetBatch (GetBatchRequest) returns (GetBatchResponse);
}

message GetBatchRequest {
  repeated TableGetBatchRequest tables = 1;
}

message TableGetBatchRequest {
  string table_name = 1;
  repeated string keys = 2;
}

message GetBatchResponse {
  repeated TableGetBatchResponse tables = 1;
}

message TableGetBatchResponse {
  string table_name = 1;
  repeated GetBatchItem items = 2;
}

message GetBatchItem {
  string key = 1;
  bytes value = 2;
}

// ORCHESTRATOR
service OrchestratorService {
  rpc ListRuns (ListRunsRequest) returns (ListRunsResponse);
  rpc DumpChangelog (DumpChangelogRequest) returns (DumpChangelogResponse);
  rpc KickOffWALCompaction (KickOffWALCompactionRequest) returns (KickOffWALCompactionResponse);
  rpc GetJobStatus (GetJobStatusRequest) returns (GetJobStatusResponse);
}

message KickOffWALCompactionRequest {
}

message KickOffWALCompactionResponse {
  int64 job_id = 1;
}

message GetJobStatusRequest {
  int64 job_id = 1;
}

message GetJobStatusResponse {
  string status = 1;
}

message ListRunsRequest {
}

message ListRunsResponse {
  repeated RunMetadata runs = 1;
}

message DumpChangelogRequest {
}

message DumpChangelogResponse {
  repeated ChangelogEntry entries = 1;
}

message ChangelogEntry {
  oneof entry {
    ChangelogEntryV1 v1 = 1;
  }
}

message ChangelogEntryV1 {
  repeated string runs_added = 1;
  repeated string runs_removed = 2;
}

message RunMetadata {
  string id = 1;

  oneof belongs_to {
    int64 wal_seq_no = 2;
    TableBuffer table_buffer = 3;
    TableTree table_tree = 4;
  };

  oneof stats {
    StatsV1 stats_v1 = 1001;
  }
}

message TableBuffer {
  string table_name = 1;
  int64 seq_no = 2;
}

message TableTree {
  string table_name = 1;
  uint64 level = 2;
}

message StatsV1 {
  string min_key = 1;
  string max_key = 2;
  uint64 size_bytes = 3;
  uint64 item_count = 4;
}